src/raftstore/store/peer.rs

```diff
@@ -935,14 +946,42 @@ impl Peer {
                 });
         }
 
-        if self.has_pending_snapshot() && !self.ready_to_handle_pending_snap() {
-            debug!(
-                "{} [apply_idx: {}, last_applying_idx: {}] is not ready to apply snapshot.",
-                self.tag,
-                self.get_store().applied_index(),
-                self.last_applying_idx
-            );
-            return;
+        if let Some(snap) = self.get_pending_snapshot() {
+            if !self.ready_to_handle_pending_snap() {
+                debug!(
+                    "{} [apply_idx: {}, last_applying_idx: {}] is not ready to apply snapshot.",
+                    self.tag,
+                    self.get_store().applied_index(),
+                    self.last_applying_idx,
+                );
+                return;
+            }
+
+            let mut snap_data = RaftSnapshotData::new();
+            snap_data
+                .merge_from_bytes(snap.get_data())
+                .unwrap_or_else(|e| {
+                    warn!("{} snap data err {:?}", self.tag, e);
+                });
+            let region = snap_data.take_region();
+            // For merge process, when applying snapshot or create new peer the stale source peer is destroyed asynchronously.
+            // So here checks whether there is any overlap, if so, wait and do not handle raft ready.
+            if let Some(r) = region_ranges
```

Comment:
What about checking `epoch.version` first?

Reply:
check overlap only if the snapshot's version is larger than now?  Seems that it is a reasonable optimization, but there is no time to test. I will address it in master and later release.